Cau A

In [501]:
import pandas as pd
import numpy as np
np.random.seed(11)
def getXY(a):
    X=np.delete(a,11,1)
    Y=a[:,11]
    return X,Y



                Mô hình
$$Y = \theta_1X_1 + \theta_2X_2 + ... + \theta_{11}X_{11}$$

In [502]:
df = pd.read_csv('wine.csv', sep=';')
tt = df.columns.to_list()
a=df.to_numpy()

X,Y=getXY(a)
def Theta_hatt(X,Y):
    A=X
    AT=np.transpose(A)
    B=Y.reshape(len(Y),1)
    T=np.dot(np.linalg.inv(np.dot(AT,A)),np.dot(AT,B))
    return T
Theta=Theta_hatt(X,Y)
for i in range (len(Theta)):
    print("Theta {}:".format(i),Theta[i])
print("Y=",end='')
for i in range (len(Theta)):
    if(Theta[i]>0):
        if i == 0:
            print('{0:.5f}*x{1}'.format(Theta[i][0],i+1),end='')
        else:print('+{0:.5f}*x{1}'.format(Theta[i][0],i+1),end='')
    else:
        print('{0:.5f}*x{1}'.format(Theta[i][0],i+1),end='')


Theta 0: [0.00592516]
Theta 1: [-1.10803754]
Theta 2: [-0.26304628]
Theta 3: [0.01532228]
Theta 4: [-1.73050274]
Theta 5: [0.00380142]
Theta 6: [-0.003899]
Theta 7: [4.33858768]
Theta 8: [-0.45853548]
Theta 9: [0.72971866]
Theta 10: [0.30885865]
Y=0.00593*x1-1.10804*x2-0.26305*x3+0.01532*x4-1.73050*x5+0.00380*x6-0.00390*x7+4.33859*x8-0.45854*x9+0.72972*x10+0.30886*x11

Cau B

                   Sai số MSE
$$MSE = \frac{1}{n}\sum_{i = 1}^{n} (y_i - \hat{y_i})^2$$

In [503]:

def MSE(Y,Y_hat):
    sum=0
    for i in range (len(Y)):
        sum+=(Y[i]-Y_hat[i])**2
    sum/=len(Y)
    return sum



                Cross Validation

In [504]:
np.random.shuffle(a)
def Y_hat(X,Y,x1):
    T=Theta_hatt(X,Y)
    Y_hatt=T[0][0]*x1
    return Y_hatt

def CrossValidation(f,column,kflod):
    p=np.array_split(a,kflod)
    err=0
    for i in range (kflod): #chon 1 tập hợp làm Test
        Y=p[i][:,11]
        X=p[i][:,column]
        for j in range(kflod): #cac tập hợp còn lại là train
            if i!=j:
                g=p[j]
                for l in range(kflod):
                    if i!=l and j!=l:
                        h=p[l]
                        g=np.concatenate((g,h),axis=0)
                X2=g[:,column]
                Y2=g[:,11]
                break;
        X2=X2.reshape(len(X2),1)
        Ymu=f(X2,Y2,X)
        err+=MSE(Y,Ymu)
    return err/kflod

K=[]
for i in range (11):
    print("{0:}:{1:}".format(i,tt[i]),CrossValidation(Y_hat,i,5))
    K.append(CrossValidation(Y_hat,i,5))
idx=K.index(min(K))
X=a[:,idx]
Y=a[:,11]
X=X.reshape(len(X),1)
print('Thuộc tính đặc trưng tốt nhất là',tt[idx])
print('Mô hình tương ứng: Y=',end='')
print('{0:.5f}x'.format(Theta_hatt(X,Y)[0][0]))




        

    
    

0:fixed acidity 1.6983631494760338
1:volatile acidity 5.136446787162993
2:citric acid 9.6281176740976
3:residual sugar 6.83288609866014
4:chlorides 8.468744241612058
5:free sulfur dioxide 10.962685424685478
6:total sulfur dioxide 12.689378550333313
7:density 0.6593232860380397
8:pH 0.7603353157096413
9:sulphates 2.3061636454000127
10:alcohol 0.5240924142165829
Thuộc tính đặc trưng tốt nhất là alcohol
Mô hình tương ứng: Y=0.54371x


<h>Câu C <h> Tìm mô hình riêng cho kết quả tốt nhất

            Mô hình
$Y_1 = \theta_0 + \theta_1 X_1^2$

In [505]:

def Y1(X,Y,x1):
    X2= np.ones_like(X[:,0])
    X2= X2.reshape(len(X2),1)
    X=X*X
    X=np.concatenate((X2,X),axis=1) #1 + Xtrain
    T=Theta_hatt(X,Y)
    Y_hat=T[0][0]+T[1][0]*x1*x1
    return Y_hat
L=[]
for i in range (11):
    err=CrossValidation(Y1,i,5)
    print("{0:}:{1:}".format(i,tt[i]),err)
    L.append(err)
idx=L.index(min(L))
print('Thuộc tính đặc trưng tốt nhất là',tt[idx],end=' ')
print("có độ sai số là {0:.5f}".format(L[idx]))

0:fixed acidity 0.6418312625300946
1:volatile acidity 0.576085764001403
2:citric acid 0.6238616079706147
3:residual sugar 0.655965131743191
4:chlorides 0.6508894837170027
5:free sulfur dioxide 0.6544237553588615
6:total sulfur dioxide 0.6443782692526723
7:density 0.6328510702087332
8:pH 0.6525073656294182
9:sulphates 0.6519818552425936
10:alcohol 0.49365440788893944
Thuộc tính đặc trưng tốt nhất là alcohol có độ sai số là 0.49365


            Mô hình
$Y_2 = \theta_0 + \theta_1 X_1^3 + \theta_2 X_2^2$

In [506]:
def Y2(X,Y,x1):
    X2= np.ones_like(X[:,0])
    X2= X2.reshape(len(X2),1)
    
    X1=X*X*X
    X3=X*X
    X=np.concatenate((X2,X1),axis=1) #1 + Xtrain
    X=np.concatenate((X,X3),axis=1)
    T=Theta_hatt(X,Y)
    Y_hat=T[0][0]+T[1][0]*x1*x1*x1+T[2][0]*x1*x1
    return Y_hat

F=[]
for i in range (11):
    err=CrossValidation(Y2,i,5)
    print("{0:}:{1:}".format(i,tt[i]),err)
    F.append(err)
idx=F.index(min(F))
print('Thuộc tính đặc trưng tốt nhất là',tt[idx],end=' ')
print("có độ sai số là {0:.5f}".format(F[idx]))


0:fixed acidity 0.6415935930784299
1:volatile acidity 0.5633400141401979
2:citric acid 0.6221079642923586
3:residual sugar 0.6548453891507485
4:chlorides 0.651748538794733
5:free sulfur dioxide 0.6536119806923144
6:total sulfur dioxide 0.6097796067665078
7:density 0.6178026338826947
8:pH 0.6535476050960611
9:sulphates 0.6018991253439439
10:alcohol 0.48883834339090476
Thuộc tính đặc trưng tốt nhất là alcohol có độ sai số là 0.48884


Sai số của mô hình $Y_1 = \theta_0 + \theta_1 X_1^2$ Là : 0.49365

Sai số của mô hình $Y_2 = \theta_0 + \theta_1 X_1^3 + \theta_2 X_2^2$ Là : 0.48884

Vậy chọn mô hình $Y_2$ là mô hình tốt nhất mà Em tìm được

